File: src/emailSender.ts
import nodemailer, { Transporter } from 'nodemailer';
import { Database } from './database';
import * as dotenv from 'dotenv';

dotenv.config();

export interface EmailConfig {
  host: string;
  port: number;
  user: string;
  password: string;
  fromEmail: string;
  fromName: string;
}

export class EmailSender {
  private transporter: Transporter | null = null;
  private db: Database;
  private config: EmailConfig;

  constructor() {
    this.db = new Database();

    this.config = {
      host: process.env.SMTP_HOST || 'smtp.gmail.com',
      port: parseInt(process.env.SMTP_PORT || '587'),
      user: process.env.SMTP_USER || '',
      password: process.env.SMTP_PASSWORD || '',
      fromEmail: process.env.FROM_EMAIL || '',
      fromName: process.env.FROM_NAME || 'Web Agency',
    };

    this.initTransporter();
  }

  private initTransporter(): void {
    if (!this.config.user || !this.config.password) {
      console.log('⚠️  Email credentials not configured');
      console.log('   Add SMTP_USER and SMTP_PASSWORD to .env file');
      return;
    }

    this.transporter = nodemailer.createTransport({
      host: this.config.host,
      port: this.config.port,
      secure: false, // true pour 465, false pour les autres ports
      auth: {
        user: this.config.user,
        pass: this.config.password,
      },
    });

    console.log('✅ Email transporter initialized');
  }

  /**
   * Vérifie la configuration email
   */
  async testConnection(): Promise<boolean> {
    if (!this.transporter) {
      console.log('❌ Email transporter not configured');
      return false;
    }

    try {
      await this.transporter.verify();
      console.log('✅ SMTP connection successful');
      return true;
    } catch (error) {
      console.error('❌ SMTP connection failed:', error);
      return false;
    }
  }

  /**
   * Envoie un email à un prospect
   */
  async sendEmail(to: string, subject: string, body: string): Promise<boolean> {
    if (!this.transporter) {
      console.log('❌ Cannot send email: transporter not configured');
      return false;
    }

    try {
      const mailOptions = {
        from: `"${this.config.fromName}" <${this.config.fromEmail}>`,
        to: to,
        subject: subject,
        text: body,
        html: body.replace(/\n/g, '<br>'), // Conversion simple en HTML
      };

      const info = await this.transporter.sendMail(mailOptions);
      console.log(`✅ Email sent to ${to}`);
      console.log(`   Message ID: ${info.messageId}`);

      return true;
    } catch (error) {
      console.error(`❌ Failed to send email to ${to}:`, error);
      return false;
    }
  }

  /**
   * Envoie des emails à tous les prospects avec un email généré
   */
  async sendAllEmails(dailyLimit: number = 50): Promise<void> {
    const prospects = await this.db.getProspects('analyzed');

    console.log(`📧 Preparing to send emails to prospects...\n`);

    if (!this.transporter) {
      console.log('❌ Email not configured. Please configure SMTP settings in .env');
      return;
    }

    // Vérifier la connexion
    const connected = await this.testConnection();
    if (!connected) {
      console.log('❌ Cannot connect to email server');
      return;
    }

    let sent = 0;
    let failed = 0;

    for (const prospect of prospects) {
      if (sent >= dailyLimit) {
        console.log(`\n⚠️  Daily limit reached (${dailyLimit} emails)`);
        break;
      }

      if (!prospect.id || !prospect.email) {
        console.log(`⏭️  Skipping ${prospect.company_name}: no email`);
        continue;
      }

      // Récupérer l'email généré précédemment
      const emailLog = await this.getEmailLog(prospect.id);

      if (!emailLog) {
        console.log(`⏭️  Skipping ${prospect.company_name}: no email generated`);
        continue;
      }

      console.log(`\n📧 Sending to: ${prospect.company_name} (${prospect.email})`);

      const success = await this.sendEmail(
        prospect.email,
        emailLog.subject,
        emailLog.body
      );

      if (success) {
        sent++;
        await this.db.updateProspectStatus(prospect.id, 'contacted');

        // Pause entre les envois pour éviter d'être marqué comme spam
        await new Promise(resolve => setTimeout(resolve, 3000));
      } else {
        failed++;
      }
    }

    console.log('\n' + '='.repeat(60));
    console.log('📊 EMAIL CAMPAIGN SUMMARY');
    console.log('='.repeat(60));
    console.log(`✅ Successfully sent: ${sent}`);
    console.log(`❌ Failed: ${failed}`);
    console.log(`📝 Total prospects: ${prospects.length}`);
    console.log('='.repeat(60) + '\n');
  }

  /**
   * Envoie un email de test
   */
  async sendTestEmail(testEmail: string): Promise<void> {
    console.log(`📧 Sending test email to ${testEmail}...\n`);

    const subject = 'Test - Système de prospection automatique';
    const body = `Bonjour,

Ceci est un email de test de votre système de prospection automatique.

Si vous recevez cet email, votre configuration SMTP est correcte ! ✅

Vous pouvez maintenant :
1. Scraper des prospects (npm run scrape)
2. Analyser leurs sites web (npm run analyze)
3. Générer des emails personnalisés (npm run generate-emails)
4. Envoyer les emails (npm run send-emails)

Bonne prospection !

---
${process.env.YOUR_COMPANY_NAME || 'Votre Entreprise'}`;

    const success = await this.sendEmail(testEmail, subject, body);

    if (success) {
      console.log('✅ Test email sent successfully!');
      console.log('   Check your inbox (and spam folder)');
    } else {
      console.log('❌ Failed to send test email');
      console.log('   Check your SMTP configuration in .env');
    }
  }

  /**
   * Récupère l'email log pour un prospect
   */
  private async getEmailLog(prospectId: number): Promise<any> {
    return new Promise((resolve, reject) => {
      this.db['db'].get(
        'SELECT * FROM email_logs WHERE prospect_id = ? ORDER BY id DESC LIMIT 1',
        [prospectId],
        (err, row) => {
          if (err) reject(err);
          else resolve(row);
        }
      );
    });
  }

  /**
   * Campagne d'envoi progressive avec suivi
   */
  async runCampaign(options: {
    dailyLimit?: number;
    batchSize?: number;
    delayBetweenBatches?: number;
  } = {}): Promise<void> {
    const {
      dailyLimit = 50,
      batchSize = 10,
      delayBetweenBatches = 300000 // 5 minutes
    } = options;

    console.log('🚀 Starting email campaign...');
    console.log(`   Daily limit: ${dailyLimit} emails`);
    console.log(`   Batch size: ${batchSize} emails`);
    console.log(`   Delay between batches: ${delayBetweenBatches / 60000} minutes\n`);

    await this.sendAllEmails(dailyLimit);

    console.log('✅ Campaign completed!');
  }

  close(): void {
    this.db.close();
  }
}

// Exécution directe du script
if (require.main === module) {
  const args = process.argv.slice(2);
  const sender = new EmailSender();

  if (args[0] === 'test') {
    // Envoyer un email de test
    const testEmail = args[1] || process.env.SMTP_USER;
    if (!testEmail) {
      console.log('❌ Please provide a test email: npm run send-emails test your@email.com');
      process.exit(1);
    }

    sender.sendTestEmail(testEmail)
      .then(() => {
        sender.close();
        process.exit(0);
      })
      .catch((error) => {
        console.error('❌ Error:', error);
        process.exit(1);
      });
  } else {
    // Lancer la campagne normale
    sender.runCampaign()
      .then(() => {
        sender.close();
        console.log('✅ Email sending completed successfully');
        process.exit(0);
      })
      .catch((error) => {
        console.error('❌ Error sending emails:', error);
        process.exit(1);
      });
  }
}

File: python/email_sender.py
"""
Module d'envoi d'emails automatique
Utilise SMTP pour envoyer les emails
"""
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import os
from dotenv import load_dotenv
import time

from database import Database

load_dotenv()


class EmailSender:
    """Gère l'envoi automatique d'emails"""

    def __init__(self):
        self.db = Database()

        self.smtp_host = os.getenv('SMTP_HOST', 'smtp.gmail.com')
        self.smtp_port = int(os.getenv('SMTP_PORT', '587'))
        self.smtp_user = os.getenv('SMTP_USER', '')
        self.smtp_password = os.getenv('SMTP_PASSWORD', '')
        self.from_email = os.getenv('FROM_EMAIL', '')
        self.from_name = os.getenv('FROM_NAME', '')

        if not self.smtp_user or not self.smtp_password:
            print('⚠️  Email credentials not configured')
            print('   Add SMTP_USER and SMTP_PASSWORD to .env file')

    def test_connection(self) -> bool:
        """Teste la connexion SMTP"""
        if not self.smtp_user or not self.smtp_password:
            print('❌ Email not configured')
            return False

        try:
            server = smtplib.SMTP(self.smtp_host, self.smtp_port)
            server.starttls()
            server.login(self.smtp_user, self.smtp_password)
            server.quit()

            print('✅ SMTP connection successful')
            return True

        except Exception as e:
            print(f'❌ SMTP connection failed: {e}')
            return False

    def send_email(self, to: str, subject: str, body: str) -> bool:
        """
        Envoie un email

        Args:
            to: Destinataire
            subject: Sujet
            body: Corps de l'email

        Returns:
            True si l'envoi réussit
        """
        if not self.smtp_user or not self.smtp_password:
            print('❌ Cannot send email: SMTP not configured')
            return False

        try:
            # Créer le message
            msg = MIMEMultipart('alternative')
            msg['From'] = f'"{self.from_name}" <{self.from_email}>'
            msg['To'] = to
            msg['Subject'] = subject

            # Ajouter le corps (texte et HTML)
            text_part = MIMEText(body, 'plain')
            html_part = MIMEText(body.replace('\n', '<br>'), 'html')

            msg.attach(text_part)
            msg.attach(html_part)

            # Envoyer
            server = smtplib.SMTP(self.smtp_host, self.smtp_port)
            server.starttls()
            server.login(self.smtp_user, self.smtp_password)
            server.send_message(msg)
            server.quit()

            print(f'✅ Email sent to {to}')
            return True

        except Exception as e:
            print(f'❌ Failed to send email to {to}: {e}')
            return False

    def send_test_email(self, test_email: str):
        """Envoie un email de test"""
        print(f'📧 Sending test email to {test_email}...\n')

        subject = 'Test - Système de prospection automatique'
        body = f"""Bonjour,

Ceci est un email de test de votre système de prospection automatique.

Si vous recevez cet email, votre configuration SMTP est correcte ! ✅

Vous pouvez maintenant :
1. Scraper des prospects (python scraper.py)
2. Trouver leurs emails (python email_finder.py)
3. Analyser leurs sites web (python analyzer.py)
4. Générer des emails personnalisés (python email_generator.py)
5. Envoyer les emails (python email_sender.py)

Bonne prospection !

---
{os.getenv('YOUR_COMPANY_NAME', 'Votre Entreprise')}"""

        success = self.send_email(test_email, subject, body)

        if success:
            print('✅ Test email sent successfully!')
            print('   Check your inbox (and spam folder)')
        else:
            print('❌ Failed to send test email')
            print('   Check your SMTP configuration in .env')

    def send_all_emails(self, daily_limit: int = 50):
        """Envoie tous les emails en attente"""
        prospects = self.db.get_prospects('analyzed')

        print(f'📧 Preparing to send emails to prospects...\n')

        if not self.smtp_user or not self.smtp_password:
            print('❌ Email not configured. Please configure SMTP settings in .env')
            return

        # Vérifier la connexion
        if not self.test_connection():
            print('❌ Cannot connect to email server')
            return

        sent = 0
        failed = 0

        for prospect in prospects:
            if sent >= daily_limit:
                print(f'\n⚠️  Daily limit reached ({daily_limit} emails)')
                break

            prospect_id = prospect['id']
            email = prospect.get('email')

            if not email:
                print(f"⏭️  Skipping {prospect['company_name']}: no email")
                continue

            # Récupérer l'email généré
            email_log = self._get_email_log(prospect_id)

            if not email_log:
                print(f"⏭️  Skipping {prospect['company_name']}: no email generated")
                continue

            print(f"\n📧 Sending to: {prospect['company_name']} ({email})")

            success = self.send_email(
                email,
                email_log['subject'],
                email_log['body']
            )

            if success:
                sent += 1
                self.db.update_prospect_status(prospect_id, 'contacted')

                # Pause entre les envois
                time.sleep(3)
            else:
                failed += 1

        print('\n' + '=' * 60)
        print('📊 EMAIL CAMPAIGN SUMMARY')
        print('=' * 60)
        print(f'✅ Successfully sent: {sent}')
        print(f'❌ Failed: {failed}')
        print(f'📝 Total prospects: {len(prospects)}')
        print('=' * 60 + '\n')

    def _get_email_log(self, prospect_id: int):
        """Récupère le dernier email log pour un prospect"""
        cursor = self.db.conn.cursor()
        cursor.execute(
            'SELECT * FROM email_logs WHERE prospect_id = ? ORDER BY id DESC LIMIT 1',
            (prospect_id,)
        )
        row = cursor.fetchone()
        return dict(row) if row else None

    def close(self):
        """Ferme les ressources"""
        self.db.close()


if __name__ == '__main__':
    import sys

    sender = EmailSender()

    try:
        if len(sys.argv) > 1 and sys.argv[1] == 'test':
            # Mode test
            test_email = sys.argv[2] if len(sys.argv) > 2 else sender.smtp_user
            if not test_email:
                print('❌ Please provide a test email: python email_sender.py test your@email.com')
                sys.exit(1)

            sender.send_test_email(test_email)

        else:
            # Mode campagne normale
            sender.send_all_emails()

    except KeyboardInterrupt:
        print('\n\n⚠️  Email sending interrupted by user')
    except Exception as e:
        print(f'\n\n❌ Error sending emails: {e}')
        import traceback
        traceback.print_exc()
    finally:
        sender.close()
        print('✅ Email sending completed')
